exonum/src/api/websockets/server.rs

```diff
@@ -0,0 +1,72 @@
+// Copyright 2018 The Exonum Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use actix::prelude::*;
+
+use rand::{self, Rng, ThreadRng};
+use std::cell::RefCell;
+use std::collections::HashMap;
+
+/// WebSocket Message for communication inside websockets part.
+#[derive(Message, Debug)]
+pub struct Message(pub String);
+
+#[derive(Message)]
+#[rtype(usize)]
+pub struct Subscribe {
+    pub addr: Recipient<Syn, Message>,
+}
+
+#[derive(Message)]
+pub struct Unsubscribe {
```

Comment:
What is the purpose of `Subscribe` and `Unsubscribe` messages and the `BlockCommitWs` actor? If I understand correctly, `WsSession` is stopped when the connection with the client ends, therefore the Actor object will drop too if it is not kept in the `HashSet`. Would removing this entire module break everything? Please research the option of removing it completely, and moving all the mechanisms for starting/stopping connection to the `WsSession` entity (although it seems that they already are there). Session = subscription, Subscribe = Start session, Unsubscribe = Stop session. Unless I'm terribly wrong, this should work just fine and be a cleaner approach overall.

Reply:
`WsSession` is the entity that creates when some user connects, and have info about this user (`id`). And `BlockCommitWs`(now `ServerWs`) is server entity that contains all info about connected users. By commit 916c5d9 `ServerWs` contains list of subscribers, and this list deleted from other places. Now `shared_api_state` contains only websockets server address.
Now we have clean architecture
`Subscribe/Unsubscribe` message sent to server when user subscribes/unsubscribes
`Broadcast` message sent to server when blocks committed, and server send `Message` with hash of the block to all subscribers.
Why this approach better:
- separate messages for different events
- cleaner communication with clients and server
- extensible solution (e.g. we need to send another info to all subscribers or/and broadcast some different data, we just create new `Message`-type and implement `Handler<ThisMessage>` for `WsServer`.